### ISSUE TITLE
Odd interaction wth | dict merging

### ISSUE THREAD

**Issue by GBeauregard** [microsoft/pyright#3098]
**Describe the bug**
With some uses of `|` and dicts, I get unexpected `|` not allowed errors that depend on the dict typing.

**To Reproduce**
```python
test_map: dict[str | None, str] = {"4": "5"} | {"a": "b"}
```

```console
Searching for source files
Found 1 source file
/home/gregory/tes2.py
  /home/gregory/tes2.py:1:35 - error: Operator "|" not supported for types "dict[str, str]" and "dict[str, str]"
    Operator "|" not supported for types "dict[str, str]" and "dict[str, str]" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
Completed in 0.585sec
```

**Expected behavior**
The code type checks, or if it's incorrect to display a not-wrong error (the operator `|` *is* supported for the two types after all)

**VS Code extension or command-line**
1.1.223 commandline

**Additional context**
If you remove the `None` it type checks, but that isn't related to the specific error being shown to the enduser.

**Comment by GBeauregard** [comment]
As an aside, is no error here with `Final` intentional?
```python
from typing import Final
test_map: Final = {"4": "5"}
test_map |= {"a": "b"}
```
I guess you can consider it like a dict update, but it seems counterintuitive.

**Comment by erictraut** [comment]
Pyright is correct to emit an error in this case, but I agree that the error message is confusing. Mypy also emits an error in this case, and its error message is slightly less confusing. I've improved the error message. This will be included in the next release.
